R/proxy.R

```diff
@@ -0,0 +1,202 @@
+      data$id,
+      error = list(code = -32600, message = "Invalid Request")
+    ))
+  }
+
+  # If we made it here, it's valid JSON
+
+  if (data$method == "initialize") {
+    res <- jsonrpc_response(data$id, capabilities())
+    cat_json(res)
+  } else if (data$method == "tools/list") {
+    res <- jsonrpc_response(
+      data$id,
+      list(
+        tools = get_all_btw_tools()
+      )
+    )
+
+    cat_json(res)
+  } else if (data$method == "tools/call") {
+    result <- forward_request(buf)
+
+    # } else if (data$method == "prompts/list") {
+    # } else if (data$method == "resources/list") {
+  } else if (is.null(data$id)) {
+    # If there is no `id` in the request, then this is a notification and the
+    # client does not expect a response.
+    if (data$method == "notifications/initialized") {
+    }
+  } else {
+    cat_json(jsonrpc_response(
+      data$id,
+      error = list(code = -32601, message = "Method not found")
+    ))
+  }
+
+  buf <- ""
+}
+
+schedule_handle_message_from_client <- function() {
+  # Schedule the callback to run when stdin (fd 0) has input.
+  later::later_fd(handle_message_from_client, readfds = 0L)
+}
+
+handle_message_from_server <- function(data) {
+  schedule_handle_message_from_server()
+
+  logcat("FROM SERVER: ", data)
+
+  # The response_text is alredy JSON, so we'll use cat() instead of cat_json()
+  cat(data, "\n", sep = "")
+}
+
+schedule_handle_message_from_server <- function() {
+  r <- nanonext::recv_aio(the$proxy_socket)
+  promises::as.promise(r)$then(handle_message_from_server)
+}
+
+forward_request <- function(data) {
+  logcat("TO SERVER: ", data)
+
+  nanonext::send_aio(the$proxy_socket, data)
+}
+
+# This process will be launched by the MCP client, so stdout/stderr aren't
+# visible. This function will log output to the `logfile` so that you can view
+# it.
+logcat <- function(x, ..., append = TRUE) {
+  log_file <- acquaint_log_file()
+  cat(x, "\n", sep = "", append = append, file = log_file)
+}
+
+cat_json <- function(x) {
+  cat(to_json(x), "\n", sep = "")
+}
+
+capabilities <- function() {
+  list(
+    protocolVersion = "2024-11-05",
+    capabilities = list(
+      # logging = named_list(),
+      prompts = named_list(
+        listChanged = FALSE
+      ),
+      resources = named_list(
+        subscribe = FALSE,
+        listChanged = FALSE
+      ),
+      tools = named_list(
+        listChanged = FALSE
+      )
+    ),
+    serverInfo = list(
+      name = "R acquaint server",
+      version = "0.0.1"
+    ),
+    instructions = "This provides information about a running R session."
+  )
+}
+
+# Hacky way of getting tools from btw
+get_all_btw_tools <- function() {
+  dummy_provider <- ellmer::Provider("dummy", "dummy", "dummy")
+
+  tools <- lapply(unname(btw:::.btw_tools), function(tool_obj) {
+    tool <- tool_obj$tool()
+
+    if (is.null(tool)) {
+      return(NULL)
+    }
+
+    inputSchema <- compact(ellmer:::as_json(dummy_provider, tool@arguments))
```

Comment:
Does ellmer need an `as_json(NULL, ToolDef)` method or something that makes this easier?

Reply:
Could there be something like `McpProvider`? The format of the tool needs to look like this: https://modelcontextprotocol.io/docs/concepts/tools#tool-definition-structure